universal-application-tool-0.0.1/app/models/Version.java

```diff
@@ -76,4 +82,36 @@ public Instant getSubmitTime() {
         .filter(q -> q.getQuestionDefinition().getName().equals(name))
         .findAny();
   }
+
+  public ImmutableList<String> getTombstonedProgramNames() {
+    if (this.tombstonedProgramNames == null) {
+      return ImmutableList.of();
+    }
+    return ImmutableList.copyOf(this.tombstonedProgramNames);
+  }
+
+  public ImmutableList<String> getTombstonedQuestionNames() {
+    if (this.tombstonedQuestionNames == null) {
+      return ImmutableList.of();
+    }
+    return ImmutableList.copyOf(this.tombstonedQuestionNames);
+  }
+
+  public boolean addTombstoneForQuestion(Question question) {
+    if (this.tombstonedQuestionNames == null) {
+      this.tombstonedQuestionNames = new ArrayList<>();
+    }
+    if (this.tombstonedQuestionNames.contains(question.getQuestionDefinition().getName())) {
```

Comment:
Suggestion (blocking): calling `contains` with the question and program name in here and outside this class looks like an implementation detail leaking out of its abstraction. Please implement methods `public boolean questionIsTombstoned(QuestionDefinition qd)` and `public boolean programIsTombstoned(ProgramDefinition pd)` that encapsulate that check.

Reply:
Done.